Contributing guide (#254)
huggingface/accelerate: CONTRIBUTING.md

<!---
Copyright 2020 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

> -->

Reviewer: shouldn't this go under `LICENSE` file instead?

# How to contribute to Accelerate?

Everyone is welcome to contribute, and we value everybody's contribution. Code
is thus not the only way to help the community. Answering questions, helping
others, reaching out and improving the documentations are immensely valuable to
the community.

It also helps us if you spread the word: reference the library from blog posts
on the awesome projects it made possible, shout out on Twitter every time it has
helped you, or simply star the repo to say "thank you".

Whichever way you choose to contribute, please be mindful to respect our
[code of conduct](https://github.com/huggingface/accelerate/blob/master/CODE_OF_CONDUCT.md).

## You can contribute in so many ways!

> There are 4 ways you can contribute to Accelerate:

Reviewer: ```suggestion
Some of the ways you can contribute to Accelerate:
```

* Fixing outstanding issues with the existing code;
* Contributing to the examples or to the documentation;
* Submitting issues related to bugs or desired new features.

> Please ping @sgugger on your contribution.

Author: Not sure if that's important given you're watching the repo.

## Submitting a new issue or feature request

Do your best to follow these guidelines when submitting an issue or a feature
request. It will make it easier for us to come back to you quickly and with good
feedback.

> ### Did you find a bug?

Reviewer: Some or most of this information could also go to an issue template. The templates also support _form_ format ([example](https://github.com/scikit-learn/scikit-learn/issues/new?assignees=&labels=Bug%2CNeeds+Triage&template=bug_report.yml)) and people reporting a bug are probably not reading this document, but they would see that form or that template.

The 🤗 Accelerate library is robust and reliable thanks to the users who notify us of
the problems they encounter. So thank you for reporting an issue.

First, we would really appreciate it if you could **make sure the bug was not
already reported** (use the search bar on Github under Issues).

Did not find it? :( So we can act quickly on it, please follow these steps:

> * Include your **OS type and version**, the versions of **Python** and **PyTorch**.

Reviewer: we found that having a `get_versions()` which returns all relevant versions is helpful. Users would just paste the output of that in the issue.

Author: Nice! That's what we use in `transformers`, with `transformers-cli env`. I agree it would be nice to port it to `accelerate`.

* A short, self-contained, code snippet that allows us to reproduce the bug in
  less than 30s;
* Provide the *full* traceback if an exception is raised.

### Do you want a new feature (that is not a model)?

> A world-class feature request addresses the following points:

Reviewer: ```suggestion
A good feature request addresses the following points:
```

1. Motivation first:
* Is it related to a problem/frustration with the library? If so, please explain
  why. Providing a code snippet that demonstrates the problem is best.
* Is it related to something you would need for a project? We'd love to hear
  about it!
* Is it something you worked on and think could benefit the community?
  Awesome! Tell us what problem it solved for you.
2. Write a *full paragraph* describing the feature;
3. Provide a **code snippet** that demonstrates its future use;
4. In case this is related to a paper, please attach a link;
5. Attach any additional information (drawings, screenshots, etc.) you think may help.

If your issue is well written we're already 80% of the way there by the time you
post it.

> ## Start contributing! (Pull Requests)

Reviewer: ```suggestion
## Submitting a pull request (PR)
```

"Start contributing" implies the other ones were not contributions.

Before writing code, we strongly advise you to search through the existing PRs or
issues to make sure that nobody is already working on the same thing. If you are
unsure, it is always a good idea to open an issue to get some feedback.

You will need basic `git` proficiency to be able to contribute to
`accelerate`. `git` is not the easiest tool to use but it has the greatest
manual. Type `git --help` in a shell and enjoy. If you prefer books, [Pro
Git](https://git-scm.com/book/en/v2) is a very good reference.

Follow these steps to start contributing:

1. Fork the [repository](https://github.com/huggingface/accelerate) by
   clicking on the 'Fork' button on the repository's page. This creates a copy of the code
   under your GitHub user account.

2. Clone your fork to your local disk, and add the base repository as a remote:

   ```bash

>    $ git clone git@github.com:<your Github handle>/accelerate.git

Reviewer: this assumes the user has their public ssh key uploaded to github, it deserves a note.

   $ cd accelerate
   $ git remote add upstream https://github.com/huggingface/accelerate.git
   ```

> 3. Create a new branch to hold your development changes:

Reviewer: ```suggestion
3. Create a new branch to hold your development changes, and do this for every new PR you work on:
```

   ```bash

>    $ git checkout -b a-descriptive-name-for-my-changes

Reviewer: either checkout from `upstream/master`, or have them sync those two first, e.g.:

- Synchronize your ``main`` branch with the ``upstream/main`` branch,
   more details on `GitHub Docs <https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork>`_:

   .. prompt:: bash $

        git checkout main
        git fetch upstream
        git merge upstream/main

   ```

   **Do not** work on the `master` branch.

> 4. Set up a development environment by running the following command in a virtual environment:

Reviewer: ```suggestion
4. Set up a development environment by running the following command in a conda or a virtual environment you've created for working on this library:
```

   ```bash

>    $ pip install -e ".[dev]"

Reviewer 2: ```suggestion
   $ pip install -e ".[quality]"
```
There is no dev install for now ;-)

   ```

   (If accelerate was already installed in the virtual environment, remove
   it with `pip uninstall accelerate` before reinstalling it in editable
   mode with the `-e` flag.)

5. Develop the features on your branch.

   As you work on the features, you should make sure that the test suite
   passes. You should run the tests impacted by your changes like this (see 
   below an explanation regarding the environment variable):

   ```bash
   $ pytest tests/<TEST_TO_RUN>.py
   ```

   You can also run the full suite with the following command.

   ```bash
   $ make test
   ```

   `accelerate` relies on `black` and `isort` to format its source code
   consistently. After you make changes, apply automatic style corrections and code verifications
   that can't be automated in one go with:

   This target is also optimized to only work with files modified by the PR you're working on.

   If you prefer to run the checks one after the other, the following command apply the
   style corrections:

   ```bash

>    $ make style

Reviewer: on a windows machine they'd need a git bash or similar to run this I think.

   ```

   `accelerate` also uses `flake8` and a few custom scripts to check for coding mistakes. Quality

>    control runs in CI, however you can also run the same checks with:

Reviewer: you could also add them as git hooks, which have helped us with users' contributions.

Author: I won't have the bandwidth right now to spin a tutorial about adding git hooks for our checks, but I agree that it would definitely have its place here. 

Feel free to push a commit directly to the `contributing-guide` branch with it, otherwise happy to do it in a follow-up PR when I have the bandwidth.

Author: If you mean including the git hooks in the repository itself, I think a lot of us in the team are averse to having them in our own setups so it would be polarising :)

Reviewer: Yeah I saw. We can follow up on that later.

   ```bash
   $ make quality
   ```

   Once you're happy with your changes, add changed files using `git add` and
   make a commit with `git commit` to record your changes locally:

   ```bash
   $ git add modified_file.py
   $ git commit
   ```

   Please write [good commit messages](https://chris.beams.io/posts/git-commit/).

   It is a good idea to sync your copy of the code with the original
   repository regularly. This way you can quickly account for changes:

   ```bash
   $ git fetch upstream

>    $ git rebase upstream/master

Reviewer: in longer PRs (time-wise), rebase can be very tricky for beginner users, merge usually makes it much easier for them, and if you're squash and merging, the merge commit wouldn't show up in your history.

   ```

   Push the changes to your account using:

   ```bash
   $ git push -u origin a-descriptive-name-for-my-changes
   ```

6. Once you are satisfied (**and the checklist below is happy too**), go to the
   webpage of your fork on GitHub. Click on 'Pull request' to send your changes
   to the project maintainers for review.

7. It's ok if maintainers ask you for changes. It happens to core contributors
   too! So everyone can see the changes in the Pull request, work in your local
   branch and push the changes to your fork. They will automatically appear in
   the pull request.

> ### Checklist

Reviewer: adding a link to a _good_ PR may help them by seeing a good example.

1. The title of your pull request should be a summary of its contribution;
2. If your pull request addresses an issue, please mention the issue number in
   the pull request description to make sure they are linked (and people
   consulting the issue know you are working on it);
3. To indicate a work in progress please prefix the title with `[WIP]`. These
   are useful to avoid duplicated work, and to differentiate it from PRs ready

>    to be merged;

Reviewer: or create a _draft_ PR?

<!---

> Copyright 2020 The HuggingFace Team. All rights reserved.

Reviewer 2: ```suggestion
Copyright 2022 The HuggingFace Team. All rights reserved.
```
